File: neet_code_150_cpp/graphs/max_area_of_island.cpp
#include <iostream>
#include <vector>

// AC
class Solution {
    public:
        int maxAreaOfIsland(std::vector<std::vector<int>> grid) {
            auto [m, n] = std::pair((int)grid.size(), (int)grid[0].size());
            auto [result, result_prev] = std::pair(0, 0);

            for(int x = 0; x < m; ++x) {
                for(int y = 0; y < n; ++y) {
                    if(grid[x][y] == 1) {
                        result_prev = result;
                        result = 0;
                        dfs(grid, x, y, result);
                    }

                    result = std::max(result, result_prev);
                }
            }

            return result;
        }

    private:
        void dfs(std::vector<std::vector<int>>& grid, int x, int y, int& result) {
            int m = grid.size();
            int n = grid[0].size();

            // 境界チェック
            if(x < 0 
                    || y < 0
                    || x >= m
                    || y >= n
                    || grid[x][y] == 0
              ) 
            {
                return;
            }

            // 面積をインクリメント
            ++result;

            // 訪問済み
            grid[x][y] = 0;
            // 十字方向の再帰的な探索
            std::vector<std::pair<int, int>> directions = {{0, 1}, {1, 0}, {0, -1} ,{-1, 0}};
            for(auto& v: directions) {
                dfs(grid, x + v.first, y + v.second, result);
            }

        }
};

// 模範解答
class SolutionAns {
    public:
        int maxAreaOfIsland(std::vector<std::vector<int>> grid) {
            int m = grid.size();
            int n = grid[0].size();

            int result = 0;

            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (grid[i][j] == 1) {
                        result = std::max(result, dfs(grid, i, j, m, n));
                    }
                }
            }

            return result;
        }

    private:
        int dfs(std::vector<std::vector<int>>& grid, int i, int j, int m, int n) {
            if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == 0) {
                return 0;
            }
            grid[i][j] = 0;

            return 1 + dfs(grid, i - 1, j, m, n) + dfs(grid, i + 1, j, m, n)
                + dfs(grid, i, j - 1, m, n) + dfs(grid, i, j + 1, m, n);
        }
};

int main(void) {
    std::vector<std::vector<int>> case_1 = {
        {0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
        {0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0},
        {0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0},
        {0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0},
        {0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0},
    };
    // => 6

    std::vector<std::vector<int>> case_2 = {{0, 0, 0, 0, 0, 0, 0, 0}};
    // => 0

    Solution s_1;
    std::cout << s_1.maxAreaOfIsland(case_1) << std::endl;
    std::cout << s_1.maxAreaOfIsland(case_2) << std::endl;

    SolutionAns s_ans;
    std::cout << s_ans.maxAreaOfIsland(case_1) << std::endl;
    std::cout << s_ans.maxAreaOfIsland(case_2) << std::endl;
}

File: neet_code_150/src/graphs/max_area_of_island.rs
// AC
struct Solution {}
impl Solution {
    pub fn max_area_of_island(mut grid: Vec<Vec<i32>>) -> i32 {
        let (m, n) = (grid.len(), grid[0].len());
        let (mut result, mut result_prev) = (0, 0);

        for x in 0..m {
            for y in 0..n {
                if grid[x][y] == 1 {
                    result_prev = result;
                    result = 0;
                    Self::dfs(&mut grid, (x as i32, y as i32), &mut result);
                }
                result = result.max(result_prev);
            }
        }

        result
    }

    fn dfs(grid: &mut Vec<Vec<i32>>, v: (i32, i32), result: &mut i32) {
        let (m, n) = (grid.len(), grid[0].len());

        // 境界チェック
        if v.0 < 0
            || v.1 < 0
            || v.0 >= m as i32
            || v.1 >= n as i32
            || grid[v.0 as usize][v.1 as usize] == 0
        {
            return;
        }

        // 面積をインクリメント
        *result += 1;

        // 訪問済み
        grid[v.0 as usize][v.1 as usize] = 0;
        // 十字方向の再帰的な探索
        let directions: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];
        for s in directions {
            let v_new = (v.0 + s.0, v.1 + s.1);
            Self::dfs(grid, v_new, result);
        }
    }
}

// 模範解答
struct SolutionAns {}
impl SolutionAns {
    pub fn max_area_of_island(grid: Vec<Vec<i32>>) -> i32 {
        let mut max_area = 0;
        let mut new_grid = grid.clone();

        for x in 0..grid.len() {
            for y in 0..grid[0].len() {
                if new_grid[x][y] == 1 {
                    max_area = max_area.max(Self::dfs(&mut new_grid, x as i32, y as i32));
                }
            }
        }

        max_area
    }

    fn dfs(grid: &mut Vec<Vec<i32>>, x: i32, y: i32) -> i32 {
        if x < 0
            || y < 0
            || x >= grid.len() as i32
            || y >= grid[0].len() as i32
            || grid[x as usize][y as usize] == 0
        {
            return 0;
        }

        grid[x as usize][y as usize] = 0;

        let mut count = 1;
        let directions: [(i32, i32); 4] = [(0, 1), (1, 0), (0, -1), (-1, 0)];

        for (add_x, add_y) in directions {
            count += Self::dfs(grid, x + add_x, y + add_y);
        }

        count
    }
}

fn main() {
    let case_1 = vec![
        vec![0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
        vec![0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
        vec![0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0],
        vec![0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0],
        vec![0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0],
        vec![0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
        vec![0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0],
        vec![0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0],
    ];
    // => 6

    let case_2 = vec![vec![0, 0, 0, 0, 0, 0, 0, 0]];
    // => 0

    println!("case_1: {}", Solution::max_area_of_island(case_1.clone()));
    println!("case_2: {}", Solution::max_area_of_island(case_2.clone()));

    println!(
        "case_1: {}",
        SolutionAns::max_area_of_island(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAns::max_area_of_island(case_2.clone())
    );
}
